source/CScriptEngine.cpp

```diff
@@ -684,79 +684,67 @@ namespace CLEO
             return {};
         }
 
-        try
+        auto fsPath = FS::path(path);
+
+        // check for virtual path root
+        enum class VPref{ None, Game, User, Script, Cleo, Modules } virtualPrefix = VPref::None;
+        if(!fsPath.empty())
         {
-            auto fsPath = FS::path(path);
+            const auto root = fsPath.begin()->string(); // first path element
+            const auto r = root.c_str();
 
-            // check for virtual path root
-            enum class VPref{ None, Game, User, Script, Cleo, Modules } virtualPrefix = VPref::None;
-            auto root = fsPath.begin();
-            if(root != fsPath.end())
-            {
-                if(*root == DIR_GAME) virtualPrefix = VPref::Game;
-                else if (*root == DIR_USER) virtualPrefix = VPref::User;
-                else if (*root == DIR_SCRIPT) virtualPrefix = VPref::Script;
-                else if (*root == DIR_CLEO) virtualPrefix = VPref::Cleo;
-                else if (*root == DIR_MODULES) virtualPrefix = VPref::Modules;
-            }
+            if(_strcmpi(r, DIR_GAME) == 0) virtualPrefix = VPref::Game;
+            else if (_strcmpi(r, DIR_USER) == 0) virtualPrefix = VPref::User;
+            else if (_strcmpi(r, DIR_SCRIPT) == 0) virtualPrefix = VPref::Script;
+            else if (_strcmpi(r, DIR_CLEO) == 0) virtualPrefix = VPref::Cleo;
+            else if (_strcmpi(r, DIR_MODULES) == 0) virtualPrefix = VPref::Modules;
+        }
 
-            // not virtual
-            if(virtualPrefix == VPref::None)
+        // not virtual
+        if(virtualPrefix == VPref::None)
+        {
+            if(fsPath.is_relative())
             {
-                if(fsPath.is_relative())
-                {
-                    if(customWorkDir != nullptr)
-                        fsPath = ResolvePath(customWorkDir) / fsPath;
-                    else
-                        fsPath = GetWorkDir() / fsPath;
-
-                    auto resolved = FS::weakly_canonical(fsPath).string();
-
-                    // ModLoader support: do not expand game dir relative paths
-                    if (resolved.find(Filepath_Root) == 0)
-                        return FS::relative(resolved, Filepath_Root).string();
-                    else
-                        return resolved;
-                }
-
-                return FS::weakly_canonical(fsPath).string();
+                if(customWorkDir != nullptr)
+                    fsPath = ResolvePath(customWorkDir) / fsPath;
+                else
+                    fsPath = GetWorkDir() / fsPath;
             }
 
-            // expand virtual paths
-            FS::path resolved;
+            auto result = fsPath.string();
+            NormalizeFilepath(result, false);
 
-            if (virtualPrefix == VPref::User) // user files location
+            // ModLoader support: keep game dir relative paths relative
+            if (result.length() > Filepath_Root.length() && // and separator

```

Comment:
what's the difference between Filepath_Root and Gta_Root_Dir_Path?

Reply:
Filepath_Root sits in CLEO core as is global std::string variable. Gta_Root_Dir_Path is just pointer to path in original game.

Reply:
why we need two? can we use one?

Reply:
True. It was added later to utils.